compiler/rustc_typeck/src/check/demand.rs

```diff
@@ -347,6 +349,13 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 }
             }
 
+            // Avoid suggesting wrapping in `NonZeroU64` and alike
+            if self.tcx.layout_scalar_valid_range(expected_adt.did())
```

Comment:
Is it possible to get `the_adt.fields[0]` and check its privacy (using [this method](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/enum.Visibility.html#method.is_accessible_from) maybe, and the def id of the fcx body) instead? Or maybe best to put this privacy logic in `compatible_variants` below.

It might also help us not suggest in case like:

```
mod inner {
  pub struct Wrapper<T>(T);
}

fn needs_wrapper(t: Wrapper<i32>) {}

fn main() {
  needs_wrapper(0);
}
```

Reply:
I was thinking of doing this, but couldn't figure how to check for field accessibility 😅 

I'll try to use [`is_accessible_from`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/enum.Visibility.html#method.is_accessible_from) for this now

Reply:
Done ;)